pkgs/racket-doc/scribblings/reference/treelists.scrbl

```diff
@@ -0,0 +1,752 @@
+term ``treelist'' is used by itself, it refers to an immutable
+treelist.
+
+An immutable or mutable treelist can be used as a single-valued
+sequence (see @secref["sequences"]). The elements of the list serve as
+elements of the sequence. See also @racket[in-treelist] and
+@racket[in-mutable-treelist].
+
+@section{Immutable Treelists}
+
+@note-lib-only[racket/treelist]
+
+@history[#:added "8.12.0.7"]
+
+
+@defproc[(treelist? [v any/c]) boolean?]{
+
+Returns @racket[#t] if @racket[v] is a @tech{treelist}, @racket[#f]
+otherwise.}
+
+@defproc[(treelist [v any/c] ...) treelist?]{
+
+Returns a @tech{treelist} with @racket[v]s as its elements in order.
+
+This operation takes @math{O(log N)} time to construct a treelist of
+@math{N} elements. Unless otherwise specified, other operations on a
+treelist of length @math{N} also take @math{O(log N)} time.
+
+@examples[
+#:eval the-eval
+(treelist 1 "a" 'apple)
+]}
+
+
+@deftogether[(
+@defproc[(treelist-empty? [tl treelist?]) boolean?]
+@defthing[empty-treelist (and/c treelist? treelist-empty?)]
+)]{
+
+A predicate and constant for a @tech{treelist} of length 0.
+
+Although every empty treelist is @racket[equal?] to
+@racket[empty-treelist], since a treelist can be chaperoned via
+@racket[chaperone-treelist], not every empty treelist is @racket[eq?]
+to @racket[empty-treelist].}
+
+
+@defproc[(treelist-length [tl treelist?]) exact-nonnegative-integer?]{
+
+Returns the number of elements in @racket[tl]. This operation takes
+@math{O(1)} time.
+
+@examples[
+#:eval the-eval
+(define items (treelist 1 "a" 'apple))
+(treelist-length items)
+]}
+
+@defproc[(treelist-ref [tl treelist?] [pos exact-nonnegative-integer?]) any/c]{
+
+Returns the @racket[i]th element of @racket[tl]. The first element is
+position @racket[0], and the last position is one less than
+@racket[(treelist-length tl)].
+
+@examples[
+#:eval the-eval
+(define items (treelist 1 "a" 'apple))
+(treelist-ref items 0)
+(treelist-ref items 2)
+(eval:error (treelist-ref items 3))
+]}
+
+
+@deftogether[(
+@defproc[(treelist-first [tl treelist?]) any/c]
+@defproc[(treelist-last [tl treelist?]) any/c]
+)]{
+
+Shorthands for using @racket[treelist-ref] to access the first or last
+element of a @tech{treelist}.
+
+@examples[
+#:eval the-eval
+(define items (treelist 1 "a" 'apple))
+(treelist-first items)
+(treelist-last items)
+]}
+
+
+@defproc[(treelist-insert [tl treelist?] [pos exact-nonnegative-integer?] [v any/c]) treelist?]{
+
+Produces a treelist like @racket[tl], except that @racket[v] is
+inserted as an element before the element at @racket[pos]. If
+@racket[pos] is @racket[(treelist-length tl)], then @racket[v] is
+added to the end of the treelist.
+
+@examples[
+#:eval the-eval
+(define items (treelist 1 "a" 'apple))
+(treelist-insert items 1 "alpha")
+(treelist-insert items 3 "alpha")
+]}
+
+
+@deftogether[(
+@defproc[(treelist-cons [tl treelist?] [v any/c]) treelist?]
+@defproc[(treelist-add [tl treelist?] [v any/c]) treelist?]
```

Comment:
I see what you mean about `add`, but treelists are better at adding to the end than the beginning, so doubling down on the obscurity of `cons` with `snoc` doesn't seem like the right idea. I think `append` is common in other languages, but obviously not a good choice here. I don't have other ideas.

Reply:
push is sometimes used in other languages for adding to the end, although often in a mutable context.

Reply:
Push is also used for adding to the beginning when using a cons-list as a stack though... what about `append1`?

Reply:
`push` is also used for adding to the end when using array as a stack... So "stack" is what matters, not "cons-list".

Reply:
> treelists are better at adding to the end than the beginning

Is this worth documenting? On the one hand, if both are O(log N), maybe not. On the other hand, sometimes lists are used where order doesn't matter, and someone used to pair-based lists would probably reach for `treelist-cons` absent other information.

Reply:
Another possible set of names:

- `concat` for `append`
- `append` for `push` / `snoc`
- `prepend` for `cons`

This diverges from the Lisp(?) tradition, but it's more aligned with the rest of PLs.

Reply:
Thanks for the ideas! A survey of some languages plus a Racket datatype (although most of these are for imperative operations):

gvector: `add`
Java: `add`
C#: `Add`
Scala: `:+` on ListBuffers (where `+:` is prepend)
Clojure: `conj`
Python: `append`
Swift: `append`
Julia: `push`
Ruby: `append` and `push`
PHP: `array_push`

I'm inclined to stick with `add`, because I think the only people who expect lists to be extended at the front will also recognize `cons`, and it's easy to remember that `add` isn't `cons`. Names like `append1` and `add-right` are precise (well, I start to have reservations about "right" and "left" to mean "beginning" and "end"), but they don't look names for functions that are intended to be used commonly. I agree about improving the docs to encourage `add`.